### ISSUE TITLE
Can't authenticate with MySQL Server 5.1

### ISSUE THREAD

**Issue by Kagamine** [mysql-net/MySqlConnector#139]
Author: @mguinness
Date: Nov 23rd
Original Issue: https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/123

----------------------------------------------

Using MySQL version 5.1.38 on Windows I get the following error, but using MySQL version 5.6.27 on Linux I do not.  Also this isn't an issue with either the official MySql or Sapient Guardian providers,  Any ideas/suggestions?

```
System.NotSupportedException: Authentication method '' is not supported.
   at MySql.Data.Serialization.MySqlSession.<ConnectAsync>d__22.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySql.Data.MySqlClient.ConnectionPool.<GetSessionAsync>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySql.Data.MySqlClient.MySqlConnection.<CreateSessionAsync>d__55.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySql.Data.MySqlClient.MySqlConnection.<OpenAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySql.Data.MySqlClient.MySqlConnection.Open()
   at Microsoft.EntityFrameworkCore.Storage.Internal.MySqlRelationalConnection.Open()
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable.Enumerator.BufferlessMoveNext(Boolean buffer)
   at Microsoft.EntityFrameworkCore.Query.QueryMethodProvider.<_ShapedQuery>d__3`1.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.MySqlQueryingEnumerable`1.MySqlEnumerator.MoveNext()
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
   at lambda_method(Closure , QueryContext )
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass20_0`1.<CompileQueryCore>b__0(QueryContext qc)
   at System.Linq.Queryable.FirstOrDefault[TSource](IQueryable`1 source, Expression`1 predicate)
   at Microsoft.EntityFrameworkCore.Internal.EntityFinder`1.Find(Object[] keyValues)
   at DataQuery.Controllers.HomeController.Index() in C:\Projects\DataQuery\src\DataQuery\Controllers\HomeController.cs:line 55
```

### Further technical details

MySQL version: 5.1.38-community
Operating system: Windows Server 2008 R2
Pomelo.EntityFrameworkCore.MySql version: 1.1.0-rtm-10003 (MySqlConnector 0.7.0)

**Comment by bgrainger** [comment]
MySQL Server 5.1 doesn't support authentication plugins, which is not handled correctly by this library. (The earliest version that has been tested for compatibility is MySQL Server 5.5.)

From [Determining Authentication Method](https://dev.mysql.com/doc/internals/en/determining-authentication-method.html):

> in MySQL 5.5 arbitrary authentication methods can be implemented by means of authentication plugins.
> 
> If client or server do not support pluggable authentication (`CLIENT_PLUGIN_AUTH` capability flag is not set) then the authentication method used is inferred from client and server capabilities as follows:
>
> * The method used is Old Password Authentication if `CLIENT_PROTOCOL_41` or `CLIENT_SECURE_CONNECTION` are not set.
> * The method used is Secure Password Authentication if both `CLIENT_PROTOCOL_41` and `CLIENT_SECURE_CONNECTION` are set but `CLIENT_PLUGIN_AUTH` is not set.

The fix will be to check `InitialHandshakePacket. ProtocolCapabilities.HasFlag(ProtocolCapabilities.PluginAuth)` [here](https://github.com/mysql-net/MySqlConnector/blob/7672a8d2477b4a4cdd053316265740961c9b6b77/src/MySqlConnector/Serialization/MySqlSession.cs#L74) and fall back to using "old" or "secure" authentication based on the `ProtocolCapabilities.SecureConnection` flag. (This would also be a good time to implement Old Password Authentication.)

**Comment by bgrainger** [comment]
Fixed in [0.7.2](https://www.nuget.org/packages/MySqlConnector/0.7.2).

**Comment by bgrainger** [comment]
> This would also be a good time to implement Old Password Authentication.

This was not implemented, as it's only necessary for pre-4.1 MySQL Servers, which doesn't seem worth supporting.

**Comment by mguinness** [comment]
Confirmed fixed, thanks to all those that contributed.
